### ISSUE TITLE
False positive for `Lint/Void`

### ISSUE THREAD

**Issue by marocchino** [rubocop/rubocop#13786]
Let's consider the following code. In this code, the last name is used as the return value, so I think it shouldn't be removed.
```rb
def name=(name)
  @name = name
  reset
  name
end
```

--------

## Expected behavior

No offence

## Actual behavior

```sh
$ echo "def name=(name); @name = name; reset; name; end" | bundle exec rubocop --stdin test.rb --only Lint/Void
Inspecting 1 file
W

Offenses:

test.rb:1:39: W: [Correctable] Lint/Void: Variable name used in void context.
def name=(name); @name = name; reset; name; end
                                      ^^^^
```

## Steps to reproduce the problem

see actual behavior.

## RuboCop version

```
$  rubocop -V
1.71.1 (using Parser 3.3.7.0, rubocop-ast 1.38.0, analyzing as Ruby 3.1, running on ruby 3.3.4) [arm64-darwin23]
  - rubocop-performance 1.23.1
  - rubocop-rake 0.6.0
```

**Comment by dvandersluis** [comment]
I don't think this is actually a bug, a setter method returns the given value, regardless of what the method itself returns, similarly to initializers:

```ruby
class Test
  def name=(name)
    @name = name
   nil
  end
  attr_reader :name
end

t = Test.new
t.name = 'test'
=> "test"
t.name
=> "test"
t
=> #<Test:0x000000015565e040 @name="test">
```

As such, assignment methods are explicitly marked as being in a void context in rubocop-ast:

https://github.com/rubocop/rubocop-ast/blob/0401cf2568fbe5e9f0ed4cec68ac42046901d94b/lib/rubocop/ast/node/def_node.rb#L12-L17

**Comment by Earlopain** [comment]
Yes, this it true. Although, nothing is stopping you from calling them manually, at which point they behave like regular methods:

```rb
class Test
  def initialize
    @name = nil
    'foo'
  end

  def name=(name)
    @name = name
    nil
  end

  attr_reader :name
end

t = Test.new
pp t.send(:name=, 'test') # => nil
pp t.name # => 'test'
pp t.send(:initialize) # => 'foo'
```

This also applies to `initialize`. I don't really think these cases should be considered.

**Comment by marocchino** [comment]
Got it.
However, Ruby's default setter usually returns the input as is, and this rule may override that behavior.
Since Lint/Void is marked as a safe autocorrection, this could alter the existing behavior. Therefore, I think moving it to the unsafe category might be a reasonable approach. What do you think?

**Comment by koic** [comment]
Sacrificing safety for corner cases is questionable. While it is ideal for `foo=` to be void, but it may break the return value, so documenting and allowing it could be better in this cop. Since this is a Lint, not a Style, its strong enforcement makes it important to prevent false positives.
